### ISSUE TITLE
SBML import stuck (probably endless loop)

### ISSUE THREAD

**Issue by matthiaskoenig** [AMICI-dev/AMICI#1314]
**What did you expect to happen?**
SBML model to import.

**What has happened instead?**
The import never finished (I stopped the python interpreter after ~15-20 min). It should not take so long to process the model. Looks like an endless loop, probably in resolving the symbol tree or something similar.

```
2020-10-21 17:49:57.719 - amici.sbml_import - INFO - Finished gathering local SBML symbols        + (6.01E-01s)
2020-10-21 17:49:57.740 - amici.sbml_import - INFO - Finished processing SBML parameters          + (2.11E-02s)
2020-10-21 17:49:57.744 - amici.sbml_import - INFO - Finished processing SBML compartments        + (2.83E-03s)
2020-10-21 17:49:57.772 - amici.sbml_import - INFO - Finished flattening SBML species initials  +++ (2.33E-04s)
2020-10-21 17:49:57.772 - amici.sbml_import - INFO - Finished processing SBML species initials   ++ (2.14E-02s)
2020-10-21 17:49:58.014 - amici.sbml_import - INFO - Finished processing SBML rate rules         ++ (2.42E-01s)
2020-10-21 17:49:58.014 - amici.sbml_import - INFO - Finished processing SBML species             + (2.70E-01s)
2020-10-21 17:49:58.253 - amici.sbml_import - INFO - Finished processing SBML reactions           + (2.39E-01s)
```
Single core is running on 100 %, memory stays constant.

**To Reproduce**
Load the example model attached (v33).
```
import amici
import logging

# SBML model we want to import
sbml_file = 'body33_flat.xml'
# Name of the model that will also be the name of the python module
model_name = 'body33_glucose'
# Directory to which the generated model code is written
model_output_dir = model_name


# Create an SbmlImporter instance for our SBML model
sbml_importer = amici.SbmlImporter(sbml_file)

sbml_importer.sbml2amici(
    model_name,
    model_output_dir,
    verbose=logging.INFO,
    # observables=observables,
    # constant_parameters=constantParameters,
    # sigmas=sigmas
)
```

**AMICI version and system environment**
 - OS and version: ubuntu 20.04
 - AMICI interface: python
 - AMICI version: 0.11.8

Example SBML model
[body33_flat.zip](https://github.com/AMICI-dev/AMICI/files/5417238/body33_flat.zip)

**Comment by FFroehlich** [comment]
Had a short look at it and it looks like it is not an infinite loop but just a very inefficient implementation that becomes problematic with models with a large number of interdependent rules, so this should finish eventually, not sure how long it may take though. 

Refactoring that part of the code was the next step that was already planned and we have most element in place to make this much more efficient.

**Comment by matthiaskoenig** [comment]
Yes, finished after 120 minutes ;). But gave the following errors: 
```
2020-10-21 17:49:57.719 - amici.sbml_import - INFO - Finished gathering local SBML symbols        + (6.01E-01s)
2020-10-21 17:49:57.740 - amici.sbml_import - INFO - Finished processing SBML parameters          + (2.11E-02s)
2020-10-21 17:49:57.744 - amici.sbml_import - INFO - Finished processing SBML compartments        + (2.83E-03s)
2020-10-21 17:49:57.772 - amici.sbml_import - INFO - Finished flattening SBML species initials  +++ (2.33E-04s)
2020-10-21 17:49:57.772 - amici.sbml_import - INFO - Finished processing SBML species initials   ++ (2.14E-02s)
2020-10-21 17:49:58.014 - amici.sbml_import - INFO - Finished processing SBML rate rules         ++ (2.42E-01s)
2020-10-21 17:49:58.014 - amici.sbml_import - INFO - Finished processing SBML species             + (2.70E-01s)
2020-10-21 17:49:58.253 - amici.sbml_import - INFO - Finished processing SBML reactions           + (2.39E-01s)
2020-10-21 19:51:06.198 - amici.sbml_import - INFO - Finished processing SBML rules               + (7.27E+03s)
2020-10-21 19:51:06.202 - amici.sbml_import - INFO - Finished processing SBML initial assignments + (3.84E-03s)
2020-10-21 19:51:06.248 - amici.sbml_import - INFO - Finished processing SBML species references  + (4.52E-02s)
2020-10-21 19:51:06.248 - amici.sbml_import - INFO - Finished importing SBML                        (7.27E+03s)
Traceback (most recent call last):
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/sympy/core/sympify.py", line 406, in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 1008, in parse_expr
    return eval_expr(code, local_dict, global_dict)
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/sympy/parsing/sympy_parser.py", line 902, in eval_expr
    expr = eval(
  File "<string>", line 1
    Float ('0.5' )*log (Integer (2 )*pi *Function ('sigmayglucose' )(Symbol ('arterial' )Symbol ('plasma' ))**Integer (2 ))+Float ('0.5' )*((Function ('yglucose' )(Symbol ('arterial' )Symbol ('plasma' ))-Function ('myglucose' )(Symbol ('arterial' )Symbol ('plasma' )))/Function ('sigmayglucose' )(Symbol ('arterial' )Symbol ('plasma' )))**Integer (2 )
                                                                                         ^
SyntaxError: invalid syntax

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mkoenig/git/amici-git/amici_example.py", line 17, in <module>
    sbml_importer.sbml2amici(
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/amici/sbml_import.py", line 323, in sbml2amici
    self._process_observables(observables, sigmas, noise_distributions)
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/amici/logging.py", line 194, in wrapper_timer
    rval = func(*args, **kwargs)
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/amici/sbml_import.py", line 1011, in _process_observables
    self.symbols[SymbolId.LLHY] = {
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/amici/sbml_import.py", line 1014, in <dictcomp>
    'value': sp.sympify(noise_distribution_to_cost_function(
  File "/home/mkoenig/envs/amici/lib/python3.8/site-packages/sympy/core/sympify.py", line 408, in sympify
    raise SympifyError('could not parse %r' % a, exc)
sympy.core.sympify.SympifyError: Sympify of expression 'could not parse '0.5*log(2*pi*sigmayglucose (arterial plasma)**2) + 0.5*((yglucose (arterial plasma) - myglucose (arterial plasma)) / sigmayglucose (arterial plasma))**2'' failed, because of exception being raised:
SyntaxError: invalid syntax (<string>, line 1)
```

**Comment by FFroehlich** [comment]
Hmm, looks like some part of the code is using names instead of ids which can contain whitespaces that lead to parsing errors. Will fix.

**Comment by FFroehlich** [comment]
All of the issues (including #1312) are now be fixed in the `refactor_sbml_rules` branch, which will slowly make its way towards master. I would recommend to specify not using any observables by passing `observables={}`
```
sbml_importer.sbml2amici(
    model_name,
    model_output_dir,
    verbose=logging.INFO,
    observables={},
    # constant_parameters=constantParameters,
    # sigmas=sigmas
)
```
as otherwise observables will be generated for all SBML variables, which is quite a few for this particular model. Without the observables, compilation finished in about 2-3 minutes, so we were able to reduce import and compilation time by quite a bit.

As soon as we have made sure that all the SBML tests pass, the PR will be merged into develop and you could start using it by installing AMICI via `pip3 install -e git+https://github.com/AMICI-dev/AMICI@develop#egg=amici\&subdirectory=python/sdist`.

**Comment by FFroehlich** [comment]
Now live on develop

**Comment by matthiaskoenig** [comment]
@FFroehlich perfect. I was not aware of the observables. When running this with the latest develop version the import finishes in ~2 min, but I get the following errors (example attached)
[amici_example.zip](https://github.com/AMICI-dev/AMICI/files/5441768/amici_example.zip)
: 
```
...
2020-10-26 22:51:38.830 - amici.ode_export - INFO - Finished simplifying total_cl               +++ (6.47E-05s)
2020-10-26 22:51:38.830 - amici.ode_export - INFO - Finished computing total_cl                  ++ (3.91E-04s)
2020-10-26 22:51:38.833 - amici.ode_export - INFO - Finished writing total_cl.cpp                 + (3.21E-03s)
2020-10-26 22:51:38.834 - amici.ode_export - INFO - Finished simplifying x_solver               +++ (4.22E-04s)
2020-10-26 22:51:38.834 - amici.ode_export - INFO - Finished computing x_solver                  ++ (9.18E-04s)
2020-10-26 22:51:38.857 - amici.ode_export - INFO - Finished writing x_solver.cpp                 + (2.44E-02s)
2020-10-26 22:51:39.022 - amici.ode_export - INFO - Finished generating cpp code                    (1.02E+02s)
Traceback (most recent call last):
  File "/home/mkoenig/git/amici-git/amici_example.py", line 20, in <module>
    sbml_importer.sbml2amici(
  File "/home/mkoenig/envs/amici/src/amici/python/sdist/amici/sbml_import.py", line 365, in sbml2amici
    exporter.compile_model()
  File "/home/mkoenig/envs/amici/src/amici/python/sdist/amici/logging.py", line 194, in wrapper_timer
    rval = func(*args, **kwargs)
  File "/home/mkoenig/envs/amici/src/amici/python/sdist/amici/ode_export.py", line 2224, in compile_model
    self._compile_c_code(compiler=self.compiler,
  File "/home/mkoenig/envs/amici/src/amici/python/sdist/amici/ode_export.py", line 2294, in _compile_c_code
    result = subprocess.run(script_args,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/home/mkoenig/envs/amici/bin/python', '/home/mkoenig/git/amici-git/body33_glucose/setup.py', '--quiet', 'build_ext', '--build-lib=/home/mkoenig/git/amici-git/body33_glucose']' returned non-zero exit status 1.
Testing SWIG executable swig4.0... SUCCEEDED.
Deprecated command line option: -modern. This option is now always on.
body33_glucose_x0.cpp: In function ‘void amici::model_body33_glucose::x0_body33_glucose(amici::realtype*, amici::realtype, const realtype*, const realtype*)’:
body33_glucose_x0.cpp:24:21: error: ‘FVre’ was not declared in this scope; did you mean ‘FVve’?
   24 |     x0[10] = 5.0*BW*FVre*Fblood*(1 - HCT);
      |                     ^~~~
      |                     FVve
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

Failed building the model extension.

Process finished with exit code 1
```

**Comment by FFroehlich** [comment]
should be fixed in #1324

**Comment by matthiaskoenig** [comment]
Thanks. Forgot to give feedback. This solves my issues.
